docs/zero-downtime-deployments.md

```diff
@@ -0,0 +1,120 @@
+
+## Understanding 502 Errors
+At a high level there are 3 scenarious in which one could observe 502 errors on an AKS cluster fronted with App Gateway and AGIC. In all of these the root cause is the delay one could observe in applying a IP address changes to the Application Gateway's backend pools.
+
+  - Scaling down a Kubernetes cluster:
+    - Kubernetes is instructed to lower the number of pod replicas (perhaps manually, or via Horizontal Pod Autoscaler, or some other mechanism)
+    - Pods are put in Terminating state, while simultaneously removed from the list of Endpoints.
+    - AGIC observes the fact that Pods + Endpoints changed and begins a config update on App Gateway
+     It takes somewhere between a second and a few minutes for a pod, or a list of the pods to be removed from App Gateway's backend -- meanwhile App Gateway still attempts to deliver traffic to terminated pods
+    - Result is occasional 502 errors
+
+  - Rolling Updates:
+      - Customer updates the version of the software (perhaps using `kubectl set image`)
+      - Kubernetes upgrades a percentage of the pods at a time. The size of the bucket is defined in [the strategy section of the Deployment spec](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#updating-a-deployment)
+         - Kubernetes adds a new pod with a new image - pod goes through the states from `ContainerCreating` to `Running`
+         - When the new pod is in Running state - Kubernetes terminates the old pod
+       - The process described above is repeated until all pods are upgraded
+
+  - Kubernetes terminates resource-starved pods (CPU, RAM etc)
+
+
+## Solution
+The solution below lowers the probability of running into a scenario where App Gateway's backend pool points to terminated pods, resulting in 502 error. The solution below does not completely remove this chance.
+
+Required configuration changes prior to [performing a rolling update](https://kubernetes.io/docs/tutorials/kubernetes-basics/update/update-intro/):
+
+
+1. Change the Pod and/or Deployment specs by adding [preStop container life-cycle hooks](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks), with a delay (sleep) of at least 90 seconds.
+Example:
+```yaml
+kind: Deployment
+metadata:
+  name: x
+  labels:
+    app: y
+spec:
+  ...
+  template:
+    ...
+    spec:
+      containers:
+      - name: ctr
+        ...
+        lifecycle:
+          preStop:
+            exec:
+              command: ["sleep","90"]
+```
+
+
+> Note: The "sleep" command assumes the container is based on Linux. For Windows containers the equivalent command is "timeout".
+
+The addition of the `preStop` container life cycle hook will:
+  - delay Kubernetes sending `SIGTERM` to the container by 90 seconds, but put the pod immediately in `Terminating` state
```

Comment:
This is just a thought: This is pretty specific to Linux containers? Not sure if there is a solution for Windows containers and we need to explicitly highlight this fact?